File: assetoptima_field/lib/features/assets/models/asset.dart
class Asset {
  final String id;
  final String assetCode;
  final String name;
  final String? description;
  final String? serialNumber;
  final String status;
  final String condition;
  final String? categoryId;
  final String? vendorId;
  final String? locationId;
  final String? currentEmployeeId;
  final DateTime? purchaseDate;
  final double? purchasePrice;
  final String? warrantyExpiry;
  final String? qrValue;
  final String? notes;
  final bool isActive;
  final DateTime createdAt;
  final DateTime? updatedAt;

  const Asset({
    required this.id,
    required this.assetCode,
    required this.name,
    this.description,
    this.serialNumber,
    required this.status,
    required this.condition,
    this.categoryId,
    this.vendorId,
    this.locationId,
    this.currentEmployeeId,
    this.purchaseDate,
    this.purchasePrice,
    this.warrantyExpiry,
    this.qrValue,
    this.notes,
    required this.isActive,
    required this.createdAt,
    this.updatedAt,
  });

  factory Asset.fromJson(Map<String, dynamic> json) => Asset(
        id: json['id'] as String,
        assetCode: json['asset_code'] as String? ?? json['id'] as String,
        name: json['name'] as String,
        description: json['description'] as String?,
        serialNumber: json['serial_number'] as String?,
        status: json['status'] as String? ?? 'unknown',
        condition: json['condition'] as String? ?? 'unknown',
        categoryId: json['category_id'] as String?,
        vendorId: json['vendor_id'] as String?,
        locationId: json['location_id'] as String?,
        currentEmployeeId: json['current_employee_id'] as String?,
        purchaseDate: json['purchase_date'] != null
            ? DateTime.tryParse(json['purchase_date'] as String)
            : null,
        purchasePrice: (json['purchase_price'] as num?)?.toDouble(),
        warrantyExpiry: json['warranty_expiry'] as String?,
        qrValue: json['qr_value'] as String?,
        notes: json['notes'] as String?,
        isActive: json['is_active'] as bool? ?? true,
        createdAt: DateTime.parse(json['created_at'] as String),
        updatedAt: json['updated_at'] != null
            ? DateTime.tryParse(json['updated_at'] as String)
            : null,
      );
}

File: backend/app/schemas/asset.py
"""Asset Pydantic schemas."""

from datetime import datetime

from pydantic import BaseModel, Field


class AssetCreate(BaseModel):
    asset_code: str = Field(..., max_length=100)
    serial_number: str | None = Field(None, max_length=200)
    name: str = Field(..., max_length=255)
    description: str | None = Field(None, max_length=2000)
    category_id: str = Field(...)
    vendor_id: str | None = None
    location_id: str | None = None
    condition: str = Field(default="good", max_length=20)
    purchase_date: str | None = None
    purchase_price: float | None = None
    warranty_expiry: str | None = None
    notes: str | None = Field(None, max_length=2000)


class AssetUpdate(BaseModel):
    asset_code: str | None = Field(None, max_length=100)
    serial_number: str | None = Field(None, max_length=200)
    name: str | None = Field(None, max_length=255)
    description: str | None = Field(None, max_length=2000)
    category_id: str | None = None
    vendor_id: str | None = None
    location_id: str | None = None
    condition: str | None = Field(None, max_length=20)
    status: str | None = Field(None, max_length=20)
    purchase_date: str | None = None
    purchase_price: float | None = None
    warranty_expiry: str | None = None
    notes: str | None = Field(None, max_length=2000)
    is_active: bool | None = None


class AssetResponse(BaseModel):
    id: str
    asset_code: str
    serial_number: str | None
    name: str
    description: str | None
    category_id: str
    vendor_id: str | None
    location_id: str | None
    current_employee_id: str | None
    status: str
    condition: str
    purchase_date: datetime | None
    purchase_price: float | None
    warranty_expiry: datetime | None
    qr_value: str
    notes: str | None
    is_active: bool
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class AssetQRResponse(BaseModel):
    qr_value: str
    asset_code: str
    name: str
    status: str

    model_config = {"from_attributes": True}
